### ISSUE TITLE
IngressRoute support for custom request timeout

### ISSUE THREAD

**Issue by gtriggiano** [heptio/contour#815]
**Describe the solution you'd like**
I'd like to be able to configure custom request timeouts per IngressRoute.

It seems from docs that you can use annotations on ingresses, but according to [this comment](https://github.com/heptio/contour/pull/655#discussion_r213334220) 

>  IngressRoutes don't really use annotations for anything

Docs regarding IngressRoute contain no reference to a _requests timeout config parameter_.

**Environment:**

- Kubernetes version: v1.10.3 (EKS)
- Cloud provider or hardware configuration: AWS

**Contour daemon set**
```yaml
apiVersion: extensions/v1beta1
kind: DaemonSet
metadata:
  labels:
    app: contour
  name: contour
  namespace: heptio-contour
spec:
  selector:
    matchLabels:
      app: contour
  template:
    metadata:
      annotations:
        prometheus.io/format: prometheus
        prometheus.io/path: /stats
        prometheus.io/port: "9001"
        prometheus.io/scrape: "true"
      creationTimestamp: null
      labels:
        app: contour
    spec:
      containers:
      - args:
        - serve
        - --incluster
        command:
        - contour
        image: gcr.io/heptio-images/contour:latest
        imagePullPolicy: Always
        name: contour
        ports:
        - containerPort: 8000
          hostPort: 8000
          name: contour
          protocol: TCP
      - args:
        - --config-path /config/contour.yaml
        - --service-cluster cluster0
        - --service-node node0
        - --log-level info
        - --v2-config-only
        command:
        - envoy
        image: docker.io/envoyproxy/envoy-alpine:v1.7.0
        name: envoy
        ports:
        - containerPort: 8080
          hostPort: 8080
          name: http
          protocol: TCP
        - containerPort: 8443
          hostPort: 8443
          name: https
          protocol: TCP
        volumeMounts:
        - mountPath: /config
          name: contour-config
      dnsPolicy: ClusterFirst
      hostNetwork: true
      initContainers:
      - args:
        - bootstrap
        - /config/contour.yaml
        command:
        - contour
        image: gcr.io/heptio-images/contour:latest
        imagePullPolicy: Always
        name: envoy-initconfig
        volumeMounts:
        - mountPath: /config
          name: contour-config
      restartPolicy: Always
      serviceAccount: contour
      serviceAccountName: contour
      volumes:
      - emptyDir: {}
        name: contour-config
```

**Comment by jipperinbham** [comment]
I'd like to help with this if it has not been started already. Here's what I was thinking:

- Some of the other annotations (i.e. `contour.heptio.com/websocket-routes`) have been added to the `Route` struct of the `IngressRouteSpec` which ends up in the [Route portion of the dag](https://github.com/heptio/contour/blob/dd412d58122c23f2236489cee32d78e65201c81d/internal/dag/dag.go#L51-L84). 
- It would seem the best place to define the retry and timeout annotations is in the [Route struct](https://github.com/heptio/contour/blob/dd412d58122c23f2236489cee32d78e65201c81d/apis/contour/v1beta1/ingressroute.go#L53-L68). 
- The one caveat being [this comment](https://github.com/heptio/contour/blob/dd412d58122c23f2236489cee32d78e65201c81d/internal/dag/dag.go#L61) and [this one](https://github.com/heptio/contour/blob/dd412d58122c23f2236489cee32d78e65201c81d/internal/dag/dag.go#L67) would suggest all of the below would need to be defined on the `Service` (both dag and IngressRoute)

Below would be the additional fields (current fields omitted for brevity):

```go
type Route struct {
  // RequestTimeout specified as a time.Duration set the upstream timeout for the route
  RequestTimeout string `json:"request-timeout,omitempty"`
  // RetryOn specifies the conditions under which retry takes place
  RetryOn string `json:"retry-on,omitempty"`
  // NumRetries defines the maximum number of retries Envoy should make before abandoning 
  // and returning an error to the client
  NumRetries int `json:"num-retries,omitempty"`
  // PerTryTimeout specificed as a time.Duration, set the timeout per retry attempt
  PreTryTimeout string `json:"per-try-timeout"`
}
```

I'm open to also take on the work of moving these things around if that's what is needed for this to be done well.

**Comment by lostllama** [comment]
To add my 2 cents to this, we're using the following annotations on Ingress that we'd like to use on IngressRoute too:

    ingress.kubernetes.io/force-ssl-redirect
    contour.heptio.com/retry-on
    contour.heptio.com/num-retries
    contour.heptio.com/request-timeout
    contour.heptio.com/per-try-timeout

**Comment by lostllama** [comment]
It seems that the fields indicated to @jipperinbham need to be added [here](https://github.com/heptio/contour/blob/b65eab04438ea7aef3fbd587cce123bf4c5cd620/apis/contour/v1beta1/ingressroute.go) and mapped through to Envoy.

I'm currently using the following against Ingress, but also need them on IngressRoute:

* ingress.kubernetes.io/force-ssl-redirect
* contour.heptio.com/retry-on
* contour.heptio.com/num-retries
*  contour.heptio.com/request-timeout
*  contour.heptio.com/per-try-timeout

**Comment by davecheney** [comment]
Thank you for your comment. We won’t be adding these annotations but they will be added as fields on the ingress route object where applicable.

**Comment by lostllama** [comment]
Understood, I just wanted to add my 2 cents for the features that I need. Thank you for your prompt response.

**Comment by rsyvarth** [comment]
Hey @davecheney would you accept a PR adding these fields to IngressRoute and hooking them up to the Route in Envoy (as @jipperinbham suggested)?
```go
	// RequestTimeoutSeconds specified as a time.Duration set the upstream timeout for the route
	RequestTimeoutSeconds int64 `json:"requestTimeoutSeconds,omitempty"`
	// RetryOn specifies the conditions under which retry takes place
	RetryOn string `json:"retryOn,omitempty"`
	// NumRetries defines the maximum number of retries Envoy should make before abandoning
	// and returning an error to the client
	NumRetries uint32 `json:"numRetries,omitempty"`
	// perTryTimeoutSeconds specificed as a time.Duration, set the timeout per retry attempt
	PreTryTimeoutSeconds int64 `json:"perTryTimeoutSeconds,omitempty"`
```

**Comment by rsyvarth** [comment]
Great! I am mostly concerned with the RetryOn/NumRetries fields personally - we see in our cluster that envoy will often attempt a request on a closed connection (if a pod gets killed for example) and which returns a 503 immediately. We reduce our error rate pretty significantly by enabling the envoy-level retries. 

I am not sure that retry logic error makes a ton of sense as a part of a generic timeout object, what do you think about just adding in these fields to the Route?
```go
	// RetryOn specifies the conditions under which retry takes place
	RetryOn string `json:"retryOn,omitempty"`
	// NumRetries defines the maximum number of retries Envoy should make before abandoning
	// and returning an error to the client
	NumRetries uint32 `json:"numRetries,omitempty"`
```
(thanks for all of the work on this project btw! Currently migrating our cluster over from Ambassador and really appreciate the k8s-native design being used here).

**Comment by prasoontelang** [comment]
I am willing to work on this if nobody is actively working. Will start with the design doc.

**Comment by davecheney** [comment]
All yours.

**Comment by unicell** [comment]
> WRT to timeouts, I don’t want to add additional fields to route, but create a new timeout object that encapsulates all the elements of timeout behaviour which we can apply in all the places where timeouts are needed—which is turning out to be a growing set.

@davecheney do you mind be more specific about the last part? Are you referring to that growing set as something in Envoy API or within Contour, like HealthCheck?

**Comment by davecheney** [comment]
Because of the need to release Contour 0.11 to address the security issue in Envoy 1.9.0, this issue has been reassigned to the 0.12 milestone.

**Comment by stevesloka** [comment]
Fixed via #1018

**Comment by davecheney** [comment]
@stevesloka #1019 is still open. We need to update the documentation for these new parameters.

**Comment by stevesloka** [comment]
Docs implemented in #1042
